docs/sources/architecture/ring.md

```diff
@@ -0,0 +1,78 @@
+---
+title: Ring
+weight: 1000
+---
+
+# Ring
+
+This document builds upon the information in the [Loki Architecture](./) page
+
+It explains what exactly is the ring, what problem does it solve and how it actually works.
+
+We use ring in Distributor and Ingester components of Loki.
+
+Loki in microservice mode usually can have multiple ingesters and distributors. This multiple instances of same component(ingester or distributor) forms a ring (more precisely Consistent Hash Ring).
+
+Both distributors and ingesters have their own ring. Write path looks like Client -> Distributors (ring) -> Ingesters (ring). Read path looks like Client -> Querier -> Ingesters (ring).
+
+## The Problem
+
+Loki aggregates incoming log lines into something called Log stream. Log stream is just a set of logs associated with a single tenant and a unique labelset(key value pairs).
+
+Here is the problem, say I have large number of log streams coming in and I have a bunch of Loki servers(can be ingesters or distributors). Now I want to distribute these log streams across the servers in a way I can find it later when I want to read back the log stream. Here is the tricky part. I want to do this without any global directory or lookup service.
+
+So basically, we are talking problem of distributing the log streams "deterministically"!
+
+Problem is simple if all I have is single Loki server. Solution is send all logs into that single server. But having multiple servers makes the problem bit tricky
+
+Another way to solve this problem is via hashing. We assign an integer to the servers (say 0, 1, 2, 3.. etc) and we hash our log stream (labelset + tenantID) to integer value 'h' and handover it to the server with the value 'h%n' where n is the max server.
+
+Interestingly, this approach solves some of the problem we have, say same log stream goes to same server (because same set of labels + tenant gives same hash value), and while reading, we can hash it back to find the server where its value is stored.
```

Comment:
> and while reading, we can hash it back to find the server where its value is stored.

I wouldn't mention this because we don't actually perform this optimization. It's very difficult to do as it _changes_ when the ring membership changes. To get around this, we query all ingesters.